### ISSUE TITLE
Increasing RTT when using rpc_async in a PS architecture setup

### ISSUE THREAD

**Issue by anj-s** [pytorch/pytorch#57158]
## 🐛 Bug

When using a PS architecture setup for aggregating tensors, I notice that the time interval between sending a rpc_async request and when the callback is triggered increases. Given that there is only 1 worker and the PS is essentially just sending back the tensor as soon as it receives it, I would expect the RTT to be the same for all rpc_async calls.

Am I missing something? Why do I see an increase in the RTT?

## To Reproduce
Run the following gist with --num_buckets --use_cuda_tensors --bucket_size(in MB)
https://gist.github.com/anj-s/87fcd2b504b29642e0fc00639e776d8e

Timing info with --num_buckets=20 --bucket_size=10 --use_cuda_tensors
https://gist.github.com/anj-s/abd393afce896476ed626cfa4dc699da

## Expected behavior

RTT of all requests should be around some interval of time.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
  python collect_env.py
```
Collecting environment information...
PyTorch version: 1.9.0.dev20210428+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.9.0.dev20210428+cu111
[pip3] torchaudio==0.9.0.dev20210428
[pip3] torchtext==0.10.0a0+34ccda8
[pip3] torchvision==0.10.0.dev20210428+cu111
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.0.221             h6bb024c_0
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.2.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] torch                     1.9.0.dev20210428+cu111           <pip>
[conda] torchaudio                0.9.0.dev20210428           <pip>
[conda] torchtext                 0.10.0a0+34ccda8           <pip>
[conda] torchvision               0.10.0.dev20210428+cu111           <pip>

**Comment by anj-s** [comment]
cc @lw

**Comment by mrzzd** [comment]
It seems if you sleep (say for 100ms) between rpc requests, the delay goes away. So I guess the problem is with the overlap of requests. I tried increasing number of workers on tensor pipe, but did not help. @lw

**Comment by anj-s** [comment]
@lw friendly ping! Did you get a chance to look into this?

**Comment by pritamdamania87** [comment]
@anj-s It looks like this is the method that you are calling over RPC: https://gist.github.com/anj-s/87fcd2b504b29642e0fc00639e776d8e#file-repro_bucket_rtts-py-L48. Since it is a python method, the execution would be single threaded due to GIL. If you can torchscript the method, I think you would see much better performance since there would be no GIL.

**Comment by rohan-varma** [comment]
Even with TorchScripting `agg_grads`, do we expect to see a significant performance improvement with the lock? Currently due to the lock it seems like only one trainer's `agg_grad` call can run at a time. If this is also the case with TorchScript, I'm not sure if we'd expect a significant speedup.

**Comment by mrshenli** [comment]
Hey @anj-s, I tried to isolate the source of the delay, and here is what I got. 

```
[Min Delay, Max Delay] - Scenario
[4.11ms, 26.4ms] - Python + comp + alloc + cat + lock (original code)
[2.92ms, 8.21ms] - Python
[3.90ms, 9.02ms] - Python + comp
[2.70ms, 7.89ms] - Script
[3.63ms, 7.52ms] - Script + comp 
[2.10ms, 5.74ms] - Script + 1-elem tensor
[0.23ms, 0.29ms] - Script + non-tensor (integer)
```

The no-compute script function looks like this

```python
@torch.jit.script
def script_agg_grads(grads: Tensor, bucket_id: int) -> Tuple[int, Tensor]:
    return (bucket_id, grads)
```

Fo the with-compute version, I just added a few `grads += grad`, as that's what's needed for the PS. I removed the Tensor allocation and `torch.cat`, because seems like `return (bucket_id, grads)` is sufficient.

Based on the above numbers, there are two major sources of delay 1) lock (threading.Lock and GIL) 2) Tensor comm. To better understand the Tensor comm delay, I measured the delay breakdown for individual RPCs. I got:

```
tensorpipeSeralize/Deserialize - around 10us
tensorpipeAllocate - [5us, 600us]
```

The 600us must be CUDACachingAllocator misses. Although they will run in different streams on the callee, I suspect they CUDA won't be able to parallelize those, as CUDA streams works best for parallelizing comp with H2D-comm/D2H-comm. Allocation might need a global sync. (Please correct me if I was wrong.) So this suggest 1 round of warmup is insufficient to fully populate blocks in the cache. Then I tried warmup 10 times before launching the real experiment. And I got:

```
1-elem Tensor: [1.36ms, 2.28ms]
Original size tensor (self.bucket_size * 2**20 // 4): [1.12ms, 2.45ms]
```

If we look at the full trace of the original size tensor experiments, the delay no longer steadily increase:

```
Callback triggered in 1.12215 ms
Callback triggered in 1.228123 ms
Callback triggered in 1.432553 ms
Callback triggered in 1.485882 ms
Callback triggered in 1.588095 ms
Callback triggered in 1.59258 ms
Callback triggered in 2.306054 ms
Callback triggered in 2.302621 ms
Callback triggered in 2.453101 ms
Callback triggered in 2.181735 ms
Callback triggered in 2.201437 ms
Callback triggered in 2.093188 ms
Callback triggered in 2.414183 ms
Callback triggered in 2.361957 ms
Callback triggered in 1.664643 ms
Callback triggered in 1.687877 ms
Callback triggered in 1.689549 ms
Callback triggered in 1.810154 ms
Callback triggered in 1.326385 ms
Callback triggered in 1.268651 ms
```

Given that, we might need to come up with a lock-free version to get the best perf.

**Comment by anj-s** [comment]
Thank you @mrshenli for digging into this! I'll work on a lock free implementation. One question I did have is how are you warming up the RPC connection? Do you send a 1 MB message 10 times? Also are you only using a TorchScript version of the non-comp function?
